### PR TITLE
refactor(apps): Adopt Nest monorepo structure to allow module sharing

### PR DESCRIPTION
**Pull request by eli-lim**
TLDR: Adopt NestJS monorepo structure for legacy-api, ocean-api, rich-list-api and other future apps / services for code sharing

Context: was working on #1060 then dove into Nest's rabbit hole pretty hard. Tried to share a module between legacy-api and ocean-api and ended up here. 😂 Fully aware that this can be throwaway work, but we can refer to it in the future if the need arises.

#### What this PR does / why we need it:
- Original motivation - sharing the `ActuatorModule` for service health checks. Right now it's being ctrl+c ctrl-v'd everywhere: #1060 
- We can't just import modules from another workspace with its own `node_modules` as Nest will end up confused and throw some strange errors. Hence developing standalone Nest modules without the [Nest monorepo structure](https://docs.nestjs.com/cli/monorepo) is not possible **afaik** (<-- emphasis here).

#### Which issue(s) does this PR fixes?:
Greatly reduce duplicate code in the apps

#### Additional comments?:
- To limit the amount of code changed from this 1 PR, let's continue consolidating duplicate code into `libs` in future smaller PRs.
- Jest should always be run from repository root, hence checked in Intellij `.run/all-tests.run.xml` - a template run config for Jest that will make this default.
- **Note**: the build output structure will change, so might have to adjust deployment code / procedure:
    | Apps & dependent packages | Apps expanded |
    |---------|------|
    |image|image|

### CODE REVIEW

**Comment by canonbrother** [comment]
> I believe because of how Nest's IOC container works, you can't just import modules from another workspace. Hence developing standalone Nest modules without the [Nest monorepo structure](https://docs.nestjs.com/cli/monorepo) is not possible afaik (<-- emphasis here).

Ordinary nestjs mono structure only 1 node_modules
hence combine the node_modules to one then the dep can be shared to each other

**Comment by kodemon** [comment]
I think as you mentioned in your descriptions we can keep the apis in the apps folder as the website is moving out as part of referenced ticket #1071 

I like the proposed structure here, so for this PR could just do

```ts
apps
  apis|apps|service|...
  lib
  website
  ...nest config
```

for now since website will move away from this folder once I get to the aforementioned ticket.

**Review by canonbrother — DISMISSED** [review]

**Comment by fuxingloh** [comment]
> I think as you mentioned in your description we can keep the apis in the apps folder as the website is moving out as part of referenced ticket #1071
> 
> I like the proposed structure here, so for this PR could just do
> 
> ```
> apps
>   apis|apps|services|...
>     legacy-api
>     ocean-api
>   libs
>     ...shared modules
>   website
>   ...config.files
> ```
> 
> for now since website will move away from this folder once I get to the aforementioned ticket.

Speaking of this PR, maybe we should move the website out first before we merge this in. Let's wait for https://github.com/DeFiCh/jellyfish/issues/1071

File: apps/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    "experimentalDecorators": true,
+    "outDir": "./dist",
+    "baseUrl": "./",
+    "paths": {
+      "@defichain-app-lib/actuator": [
+        "libs/actuator/src"
+      ],
+      "@defichain-app-lib/actuator/*": [
+        "libs/actuator/src/*"
+      ],
```
Comment (kodemon) [inline apps/tsconfig.json:20]:
```suggestion
      "@defichain-app-lib/*": [
        "libs/*/src"
      ],
```
Wonder if we want to keep the simplicity of maintaining this file like we do with shared packages. Since this is within the scope of shared packages this space will dynamically grow as we abstract out shared functionality.

File: tsconfig.json [hunk 2]
```diff
@@ -3,6 +3,12 @@
   "compilerOptions": {
     "baseUrl": "./",
     "paths": {
+      "@defichain-app-lib/actuator": [
+        "apps/libs/actuator/src"
+      ],
+      "@defichain-app-lib/actuator/*": [
+        "apps/libs/actuator/src/*"
+      ],
```
Comment (kodemon) [inline tsconfig.json:11]:
```suggestion
      "@defichain-app-lib/*": [
        "apps/libs/*/src"
      ],
```

Comment (eli-lim) [reply tsconfig.json:11]: image

Because jest is run at the project root, it works with `tsconfig.json` to figure out where this `@defichain-app-lib/actuator` actually is. Hence I don't think removing it here is an option.

File: apps/tsconfig.json (hunk 1 above)
Comment (eli-lim) [reply apps/tsconfig.json:20]: Yea I fully agree

But removing this causes `build` to fail - typescript needs to know how to resolve the imports
image

I was also trying to figure out how best to avoid duplicating this across `jellyfish/tsconfig.json` and `jellyfish/apps/tsconfig.json`, but couldn't find an elegant way so far 🤔

Comment (eli-lim) [reply apps/tsconfig.json:20]: Woopsies - solution is for jellyfish/apps/tsconfig.json to extend jellyfish/tsconfig.json

Comment (eli-lim) [reply apps/tsconfig.json]: Generalised to `'@defichain-app-lib/(.*)': '<rootDir>/apps/libs/$1/src'`
Thanks!

**Comment by kodemon** [comment]
Added pull request for website move #1086

**Comment by fuxingloh** [comment]
FYI conflicts due to #1086 merge

**Comment by eli-lim** [comment]
All good

**Review by kodemon — APPROVED** [review]

**Review by canonbrother — APPROVED** [review]

File: lerna.json [hunk 3]
```diff
@@ -2,7 +2,8 @@
   "version": "0.0.0",
   "packages": [
     "./",
-    "apps/*",
+    "apps",
+    "apps/website",
```
Comment (fuxingloh) [inline lerna.json:6]: I believe this has been moved out, we can remove it?

File: apps/package.json [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  "license": "MIT",
+  "scripts": {
+    "build": "nest build",
+    "start:legacy-api": "nest start legacy-api",
+    "start:ocean-api": "nest start ocean-api",
+    "start:rich-list-api": "nest start rich-list-api",
+    "clean": "rm -rf dist",
+    "start:dev": "nest start --watch",
+    "start:debug": "nest start --debug --watch",
+    "start:prod": "node dist/main"
```
Comment (fuxingloh) [inline apps/package.json:17]: these scripts would not be applicable since it's mono?

**Comment by fuxingloh** [comment]
@kodemon I assume this won't break the build process for docker?

File: .run/all-tests.run.xml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<component name="ProjectRunConfigurationManager">
```
Comment (fuxingloh) [inline .run/all-tests.run.xml:1]: Is it possible to have this be placed into the `.idea` directory since it's contextually a feature in IntelliJ?

File: apps/package.json (hunk 4 above)
Comment (ivan-zynesis) [reply apps/package.json:17]: Worked for me, I did similar when testing the rich-list-api but do not find it clean, would like to hear better solution.

File: tsconfig.json [hunk 6]
```diff
@@ -3,6 +3,9 @@
   "compilerOptions": {
     "baseUrl": "./",
     "paths": {
+      "@defichain-app-lib/*": [
+        "apps/libs/*/src"
+      ],
```
Comment (fuxingloh) [inline tsconfig.json:8]:
```suggestion
      "@defichain-apps/libs-*": [
        "apps/libs/*/src"
      ],
```

Probably better to structure the import like this? To keep everything contextually related within the same org prefix.

Also; we are using organization prefix and `@defichain-app-lib` is not reserved while [`@defichain-apps`](https://www.npmjs.com/settings/defichain-apps/packages) is.

File: apps/package.json (hunk 4 above)
Comment (fuxingloh) [reply apps/package.json:17]: Since we have `start:*` e.g. `start:legacy-api`, we might not need these?

> Personally, when I develop, I never use `start:*`, I always set up tests to test my hypothesis.

**Comment by kodemon** [comment]
> @kodemon I assume this won't break the build process for docker?

That is a good question, I will take a closer look at it to confirm.

File: tsconfig.json (hunk 6 above)
Comment (eli-lim) [reply tsconfig.json:8]: Great idea thanks

**Review by fuxingloh — APPROVED** [review]

File: .run/all-tests.run.xml (hunk 5 above)
Comment (fuxingloh) [reply .run/all-tests.run.xml]: I assume this file got ignored by `.gitignore`, might need to `!` to unignore that section

File: jest.config.js [hunk 7]
```diff
@@ -3,7 +3,10 @@ module.exports = {
   testRegex: '((\\.|/)(e2e|test|spec))\\.[jt]sx?$',
   testSequencer: require.resolve('./jest.sequencer'),
   moduleNameMapper: {
-    '@defichain/(?!whale-api-client)(.*)': '<rootDir>/packages/$1/src'
+    '@defichain/(?!whale-api-client)(.*)': '<rootDir>/packages/$1/src',
+
+    // apps
```
Comment (ivan-zynesis) [inline jest.config.js:8]:
```suggestion
    // libs
```
Just to be clear non of the `app` component should be importable

**Review by ivan-zynesis — APPROVED** [review]

File: .run/all-tests.run.xml (hunk 5 above)
Comment (fuxingloh) [reply .run/all-tests.run.xml]: For this, open another PR instead after this get merged

Comment (eli-lim) [reply .run/all-tests.run.xml]: Missed that, thanks for catch

**Review by canonbrother — APPROVED** [review]
